R/faceted_qc_plot.R

```diff
@@ -1,84 +1,244 @@
+  } else {
+    shape_levels <- character()
+  }
+
+  if (plot_dose_data) {
+    colour_levels <- as.character(sort(unique(data_dose[[colour_var]])))
+  } else {
+    colour_levels <- character()
+  }
+
+  # Select variables to include in the plotly tooltips
+  tooltip_vars <- c(y_var, grouping_vars, other_tooltip_vars, x_var, colour_var)
+
+  # Preprocess the data
+  plot_data_list <- list()
+  if (plot_conc_data) {
+    plot_data_list$conc <- data_conc %>%
+      mutate(
+        legend_group = as.character(!!sym(shape_var)),
+        tooltip_text = generate_tooltip_text(., labels_df, tooltip_vars, "ADPC")
+      )
+  }
+  if (plot_dose_data) {
+    plot_data_list$dose <- data_dose %>%
+      mutate(
+        legend_group = as.character(!!sym(colour_var)),
+        tooltip_text = generate_tooltip_text(., labels_df, tooltip_vars, "ADPC")
+      )
+  }
+
+  # Return an empty plot for empty datasets
+  if (length(plot_data_list) == 0)
+    return(ggplot() + labs(title = "No data to display."))
+
+  # Combine data and create tooltip text
+  all_legend_levels <- unique(c(shape_levels, colour_levels))
+  processed_data <- bind_rows(plot_data_list) %>%
     mutate(
-      colour_factored = as.factor(!!sym(colour_var)),
-      tooltip_text = generate_tooltip_text(., labels_df, tooltip_vars, "ADPC")
+      legend_group = factor(legend_group, levels = all_legend_levels),
+      facet_title = pmap_chr(
+        select(., all_of(grouping_vars)),
+        ~ paste(list(...), collapse = ", ")
+      )
     )
 
-  plt <- ggplot(processed_data,
-                aes(x = !!sym(x_var),
-                    y = !!sym(y_var),
-                    colour = colour_factored,
-                    text = tooltip_text)) +
-    geom_point(size = 1.5) +
-    facet_grid(rows = vars(!!!syms(grouping_vars)), scales = "free_y", space = "free_y") +
+  # Define shapes
+  shape_values <- setNames(
+    c(
+      # select specific shapes for PK samples
+      rep(c(1, 4, 5, 0, 2, 6, 3), length.out = length(shape_levels)),
+      # select a filled-in circle for doses
+      rep(16, length(colour_levels))
+    ),
+    all_legend_levels
+  )
+
+  # Define colors: black for PK samples, hue palette for doses
+  if (length(colour_levels) > 0) {
+    dose_colours <- scales::hue_pal()(length(colour_levels))
+  } else {
+    dose_colours <- character()
+  }
+  colour_values <- setNames(c(rep("black", length(shape_levels)), dose_colours), all_legend_levels)
+
+  # If unique and available, format units to a string for plot labels
+  x_unit_lab <- processed_data %>% format_unit_string(x_var_units)
+  colour_unit_lab <- processed_data %>% format_unit_string(colour_var_units)
+
+  # Define a title for the legend
+  legend_title <- paste(
+    paste(
+      ifelse(
+        plot_conc_data,
+        get_label(labels_df, shape_var, "ADPC"),
+        ""
+      ),
+      ifelse(
+        plot_dose_data,
+        paste0(get_label(labels_df, colour_var, "ADPC"), colour_unit_lab),
+        ""
+      ),
```

Comment:
nitpick: There is no need to use `ifelse`, if the data and condition is not a vector:
```suggestion
      if (plot_conc_data) get_label(labels_df, shape_var, "ADPC") else "",
      if (plot_dose_data) paste0(get_label(labels_df, colour_var, "ADPC"), colour_unit_lab) else "",
```